File: native/tools/build_catalog.py
#!/usr/bin/env python3
from __future__ import annotations
import argparse, base64, hashlib, io, json, re, shutil, unicodedata
from pathlib import Path
import json5
from PIL import Image

SETS = {
    "ninja": ("Roobkaruto", "L’Ombre des Ninjas", 0xFFE07826),
    "emerald": ("Green Bafo", "Volonté Émeraude", 0xFF23A75C),
    "cod": ("Call of Trikuss", "Opérations tactiques", 0xFFB99038),
    "dbz": ("Trika Ball Z", "Puissance sans limite", 0xFFE56B25),
}

def balanced(text: str, start: int) -> str:
    opening=text[start]; closing={"[":"]","{":"}"}[opening]; depth=0; quote=None; esc=False
    for i in range(start,len(text)):
        c=text[i]
        if quote:
            if esc: esc=False
            elif c=="\\": esc=True
            elif c==quote: quote=None
            continue
        if c in "'\"`": quote=c; continue
        if c==opening: depth+=1
        elif c==closing:
            depth-=1
            if depth==0: return text[start:i+1]
    raise ValueError("Unbalanced expression")

def const_expr(text: str, name: str):
    m=re.search(rf"const\s+{re.escape(name)}\s*=\s*", text)
    if not m: return None
    i=m.end()
    while text[i].isspace(): i+=1
    return json5.loads(balanced(text,i))

def slug(s):
    s=unicodedata.normalize("NFKD",s).encode("ascii","ignore").decode().lower()
    return re.sub(r"[^a-z0-9]+","_",s).strip("_")

def decode_data(uri):
    m=re.match(r"data:([^;]+);base64,(.*)",uri,re.S)
    if not m: return None
    mime, raw=m.groups(); ext={"image/webp":"webp","image/png":"png","image/jpeg":"jpg","audio/mpeg":"mp3","audio/wav":"wav"}.get(mime,"bin")
    return ext, base64.b64decode(raw)

def save_image(uri, full: Path, thumb: Path):
    dec=decode_data(uri)
    if not dec: return False
    ext,data=dec; full=full.with_suffix('.'+ext); thumb=thumb.with_suffix('.webp')
    full.parent.mkdir(parents=True,exist_ok=True); thumb.parent.mkdir(parents=True,exist_ok=True)
    full.write_bytes(data)
    try:
        im=Image.open(io.BytesIO(data)).convert('RGB'); im.thumbnail((360,504),Image.Resampling.LANCZOS); im.save(thumb,'WEBP',quality=82,method=6)
    except Exception: thumb.write_bytes(data)
    return full.name,thumb.name

def attacks(c):
    out=[]
    for key in ('a1','a2'):
        a=c.get(key)
        if isinstance(a,list) and a: out.append({"name":str(a[0]),"damage":int(a[1] or 0) if len(a)>1 else 0,"cost":int(a[2] or 0) if len(a)>2 else 0})
    return out

def main():
    ap=argparse.ArgumentParser(); ap.add_argument('--html',required=True); ap.add_argument('--project',required=True); args=ap.parse_args()
    text=Path(args.html).read_text('utf-8',errors='ignore'); project=Path(args.project)
    assets=project/'app/src/main/assets'; docs=project/'docs'; docs.mkdir(parents=True,exist_ok=True)
    for d in ['catalog','cards/full','cards/thumb','boosters','backgrounds','ui','audio']:
        (assets/d).mkdir(parents=True,exist_ok=True)
    base=const_expr(text,'CARD_DB') or []; new=const_expr(text,'NEW_CARDS') or []; visuals=const_expr(text,'VISUALS') or {}; variants=const_expr(text,'NEW_VARIANTS') or {}; packs=const_expr(text,'OFFICIAL_PACKS') or {}; ui=const_expr(text,'UI') or {}
    merged={}
    for c in base+new:
        key=f"{c.get('set','unknown')}:{c.get('num',c.get('id','000'))}:{slug(c.get('name','card'))}"
        merged[key]=c
    cards=[]; asset_rows=[]
    for canonical,c in sorted(merged.items(),key=lambda kv:(kv[1].get('set',''),str(kv[1].get('num','')))):
        name=c.get('name','Carte'); images=variants.get(name) or ([visuals[name]] if visuals.get(name) else [])
        var_rows=[]
        for idx,uri in enumerate(images):
            if not isinstance(uri,str) or not uri.startswith('data:'): continue
            base_name=f"{slug(canonical)}__v{idx+1}"
            result=save_image(uri,assets/'cards/full'/base_name,assets/'cards/thumb'/base_name)
            if result:
                full_name,thumb_name=result; var_rows.append({"variantId":f"{canonical}:v{idx+1}","fullPath":f"cards/full/{full_name}","thumbPath":f"cards/thumb/{thumb_name}"})
        cards.append({"canonicalId":canonical,"setId":c.get('set','unknown'),"number":str(c.get('num','')),"name":name,"kind":c.get('kind','pokemon'),"stage":c.get('stage','base'),"evolvesFrom":c.get('from'),"hp":int(c.get('hp',0) or 0),"retreat":int(c.get('retreat',0) or 0),"rarity":c.get('rarity','C'),"attacks":attacks(c),"effect":c.get('effect'),"variants":var_rows})
        asset_rows.append((canonical,name,len(var_rows)))
    extensions=[]
    for sid,(name,subtitle,accent) in SETS.items():
        uri=packs.get(sid); booster=""
        if isinstance(uri,str) and uri.startswith('data:'):
            dec=decode_data(uri)
            if dec:
                ext,data=dec; p=assets/'boosters'/f'{sid}.{ext}'; p.write_bytes(data); booster=f'boosters/{p.name}'
        extensions.append({"id":sid,"name":name,"subtitle":subtitle,"accent":accent,"boosterPath":booster,"cardCount":sum(1 for c in cards if c['setId']==sid)})
    for key,value in ui.items():
        if isinstance(value,str) and value.startswith('data:image'):
            dec=decode_data(value)
            if dec:
                ext,data=dec; (assets/'ui'/f'{slug(key)}.{ext}').write_bytes(data)
    (assets/'catalog/cards.json').write_text(json.dumps(cards,ensure_ascii=False,indent=2),'utf-8')
    (assets/'catalog/extensions.json').write_text(json.dumps(extensions,ensure_ascii=False,indent=2),'utf-8')
    inventory=['# Asset Inventory','',f'- Cartes canoniques : **{len(cards)}**',f'- Variantes illustrées extraites : **{sum(x[2] for x in asset_rows)}**',f'- Boosters officiels : **{sum(1 for e in extensions if e["boosterPath"])}**','', '## Extensions']
    inventory += [f'- {e["name"]}: {e["cardCount"]} cartes — `{e["boosterPath"] or "booster manquant"}`' for e in extensions]
    inventory += ['', '## Règle de canonicalisation', '`extension:numéro:nom_normalisé` ; chaque illustration possède son propre `variantId`.']
    (docs/'ASSET_INVENTORY.md').write_text('\n'.join(inventory),'utf-8')
    (docs/'CARD_CATALOG.md').write_text('# Card Catalog\n\n'+ '\n'.join(f'- `{c["canonicalId"]}` — {c["name"]} — {c["rarity"]} — {len(c["variants"])} variante(s)' for c in cards),'utf-8')
    print(json.dumps({"cards":len(cards),"variants":sum(len(c['variants']) for c in cards),"extensions":extensions},ensure_ascii=False,indent=2))
if __name__=='__main__': main()

File: native/tools/build_catalog.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import vm from 'node:vm';

const CATALOG_SCHEMA_VERSION = 1;
const VALID_RARITIES = new Set(['C', 'U', 'R', 'SR', 'UR']);
const MAX_ASSET_BYTES = 8 * 1024 * 1024;

const args = Object.fromEntries(process.argv.slice(2).reduce((a, v, i, arr) => {
  if (v.startsWith('--')) a.push([v.slice(2), arr[i + 1]]);
  return a;
}, []));
if (!args.html || !args.project) throw new Error('Usage: --html FILE --project DIR');

const text = fs.readFileSync(args.html, 'utf8');
const project = path.resolve(args.project);
const assets = path.join(project, 'app/src/main/assets');
const catalogDir = path.join(assets, 'catalog');
const docs = path.join(project, 'docs');
for (const d of ['catalog', 'cards/full', 'cards/thumb', 'boosters', 'backgrounds', 'ui', 'audio']) {
  fs.mkdirSync(path.join(assets, d), { recursive: true });
}
fs.mkdirSync(docs, { recursive: true });

const errors = [];
const warnings = [];
const error = (code, message, context = {}) => errors.push({ severity: 'error', code, message, context });
const warn = (code, message, context = {}) => warnings.push({ severity: 'warning', code, message, context });

function balanced(start) {
  const opening = text[start], closing = opening === '[' ? ']' : '}';
  let depth = 0, quote = null, esc = false;
  for (let i = start; i < text.length; i++) {
    const c = text[i];
    if (quote) {
      if (esc) esc = false;
      else if (c === '\\') esc = true;
      else if (c === quote) quote = null;
      continue;
    }
    if (c === '"' || c === "'" || c === '`') { quote = c; continue; }
    if (c === opening) depth++;
    else if (c === closing && --depth === 0) return text.slice(start, i + 1);
  }
  throw new Error('Unbalanced expression');
}

function readConst(name) {
  const re = new RegExp(`const\\s+${name}\\s*=\\s*`, 'g');
  const m = re.exec(text);
  if (!m) return null;
  let i = m.index + m[0].length;
  while (/\s/.test(text[i])) i++;
  return vm.runInNewContext(`(${balanced(i)})`, Object.create(null), { timeout: 10000 });
}

function readJsonIfExists(file, fallback) {
  try { return JSON.parse(fs.readFileSync(file, 'utf8')); }
  catch { return fallback; }
}

function slug(s) {
  return String(s).normalize('NFKD').replace(/[\u0300-\u036f]/g, '').toLowerCase()
    .replace(/[^a-z0-9]+/g, '_').replace(/^_+|_+$/g, '');
}

function decode(uri) {
  const m = /^data:([^;]+);base64,(.*)$/s.exec(uri || '');
  if (!m) return null;
  const ext = {
    'image/webp': 'webp', 'image/png': 'png', 'image/jpeg': 'jpg',
    'audio/mpeg': 'mp3', 'audio/wav': 'wav',
  }[m[1]] || 'bin';
  return { ext, data: Buffer.from(m[2], 'base64') };
}

function saveData(uri, relBase) {
  const d = decode(uri);
  if (!d) return '';
  const rel = `${relBase}.${d.ext}`;
  const p = path.join(assets, rel);
  fs.mkdirSync(path.dirname(p), { recursive: true });
  fs.writeFileSync(p, d.data);
  return rel;
}

function attacks(c) {
  return ['a1', 'a2'].flatMap(k => Array.isArray(c[k]) ? [{
    name: String(c[k][0] || ''),
    damage: Number(c[k][1] || 0),
    cost: Number(c[k][2] || 0),
  }] : []);
}

function identityKey(setId, number) {
  return `${String(setId || '').trim()}:${String(number || '').trim()}`;
}

function newPermanentId(setId, number) {
  const set = slug(setId || 'unknown').toUpperCase() || 'UNKNOWN';
  const num = slug(number || '000').toUpperCase() || '000';
  return `CARD_${set}_${num}`;
}

const previouslyPublishedCards = readJsonIfExists(path.join(catalogDir, 'cards.json'), []);
const frozenIds = new Map();
for (const card of previouslyPublishedCards) {
  if (!card?.canonicalId || !card?.setId || !card?.number) continue;
  frozenIds.set(identityKey(card.setId, card.number), card.canonicalId);
}

const base = readConst('CARD_DB') || [];
const added = readConst('NEW_CARDS') || [];
const visuals = readConst('VISUALS') || {};
const variants = readConst('NEW_VARIANTS') || {};
const packs = readConst('OFFICIAL_PACKS') || {};
const ui = readConst('UI') || {};
const sourceCards = [...base, ...added];

const sourceByIdentity = new Map();
for (const c of sourceCards) {
  const setId = String(c.set || '').trim();
  const number = String(c.num || c.id || '').trim();
  const name = String(c.name || '').trim();
  const key = identityKey(setId, number);

  if (!setId) error('MISSING_SET_ID', 'Carte sans extension.', { name, number });
  if (!number) error('MISSING_CARD_NUMBER', 'Carte sans numéro.', { setId, name });
  if (!name) error('MISSING_CARD_NAME', 'Carte sans nom.', { setId, number });
  if (sourceByIdentity.has(key)) {
    error('DUPLICATE_CARD_NUMBER', 'Deux cartes utilisent le même numéro dans la même extension.', {
      key,
      first: sourceByIdentity.get(key)?.name,
      second: name,
    });
  } else {
    sourceByIdentity.set(key, c);
  }
}

const cards = [];
const canonicalIds = new Set();
for (const c of sourceCards.sort((a, b) => `${a.set}${a.num}`.localeCompare(`${b.set}${b.num}`))) {
  const setId = String(c.set || 'unknown');
  const number = String(c.num || c.id || '');
  const name = String(c.name || 'Carte');
  const key = identityKey(setId, number);
  const canonicalId = frozenIds.get(key) || newPermanentId(setId, number);

  if (canonicalIds.has(canonicalId)) {
    error('DUPLICATE_CARD_ID', 'CARD_ID dupliqué.', { canonicalId, setId, number, name });
  }
  canonicalIds.add(canonicalId);

  const rarity = String(c.rarity || 'C').toUpperCase();
  if (!VALID_RARITIES.has(rarity)) {
    error('UNKNOWN_RARITY', 'Rareté inconnue.', { canonicalId, rarity });
  }

  const hp = Number(c.hp || 0);
  const retreat = Number(c.retreat || 0);
  if (!Number.isFinite(hp) || hp < 0) error('INVALID_HP', 'PV invalides.', { canonicalId, hp: c.hp });
  if (!Number.isFinite(retreat) || retreat < 0) error('INVALID_RETREAT', 'Coût de retraite invalide.', { canonicalId, retreat: c.retreat });

  const cardAttacks = attacks(c);
  cardAttacks.forEach((attack, index) => {
    if (!attack.name.trim()) error('INVALID_ATTACK', 'Attaque sans nom.', { canonicalId, index });
    if (!Number.isFinite(attack.damage) || attack.damage < 0) error('INVALID_ATTACK_DAMAGE', 'Dégâts invalides.', { canonicalId, attack });
    if (!Number.isFinite(attack.cost) || attack.cost < 0) error('INVALID_ATTACK_COST', 'Coût d’attaque invalide.', { canonicalId, attack });
  });

  const imgs = variants[c.name] || (visuals[c.name] ? [visuals[c.name]] : []);
  const vars = [];
  imgs.forEach((uri, index) => {
    if (typeof uri !== 'string' || !uri.startsWith('data:')) return;
    const baseName = `${slug(canonicalId)}__v${index + 1}`;
    const full = saveData(uri, `cards/full/${baseName}`);
    const thumb = saveData(uri, `cards/thumb/${baseName}`);
    if (full) vars.push({ variantId: `${canonicalId}:v${index + 1}`, fullPath: full, thumbPath: thumb });
  });
  if (vars.length === 0) warn('MISSING_ILLUSTRATION', 'Aucune illustration intégrée pour cette carte.', { canonicalId, name });

  cards.push({
    canonicalId,
    setId,
    number,
    name,
    kind: c.kind || 'pokemon',
    stage: c.stage || 'base',
    evolvesFrom: c.from || null,
    hp,
    retreat,
    rarity,
    attacks: cardAttacks,
    effect: c.effect || null,
    variants: vars,
  });
}

const setMeta = {
  ninja: ['Roobkaruto', 'L’Ombre des Ninjas', 0xFFE07826],
  emerald: ['Green Bafo', 'Volonté Émeraude', 0xFF23A75C],
  cod: ['Call of Trikuss', 'Opérations tactiques', 0xFFB99038],
  dbz: ['Trika Ball Z', 'Puissance sans limite', 0xFFE56B25],
};

for (const card of cards) {
  if (!setMeta[card.setId]) {
    error('CARD_WITHOUT_EXTENSION', 'Carte non intégrée à une extension connue.', {
      canonicalId: card.canonicalId,
      setId: card.setId,
    });
  }
}

const extensions = Object.entries(setMeta).map(([id, [name, subtitle, accent]]) => {
  const boosterPath = saveData(packs[id], `boosters/${id}`);
  if (!boosterPath) warn('MISSING_BOOSTER_ASSET', 'Booster sans illustration.', { setId: id });
  return { id, name, subtitle, accent, boosterPath, cardCount: cards.filter(c => c.setId === id).length };
});

const namesBySet = new Map();
for (const card of cards) {
  const setNames = namesBySet.get(card.setId) || new Map();
  setNames.set(card.name, card);
  namesBySet.set(card.setId, setNames);
}
for (const card of cards) {
  if (card.evolvesFrom && !namesBySet.get(card.setId)?.has(card.evolvesFrom)) {
    warn('MISSING_EVOLUTION_SOURCE', 'Évolution référencée mais carte source introuvable dans la même extension.', {
      canonicalId: card.canonicalId,
      evolvesFrom: card.evolvesFrom,
    });
  }
}

for (const [setId, setNames] of namesBySet) {
  for (const card of setNames.values()) {
    const visited = new Set();
    let current = card;
    while (current?.evolvesFrom) {
      if (visited.has(current.name)) {
        error('CYCLIC_EVOLUTION', 'Cycle détecté dans une lignée d’évolution.', {
          setId,
          card: card.name,
          cycleAt: current.name,
        });
        break;
      }
      visited.add(current.name);
      current = setNames.get(current.evolvesFrom);
    }
  }
}

function walkFiles(dir) {
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir, { withFileTypes: true }).flatMap(entry => {
    const p = path.join(dir, entry.name);
    return entry.isDirectory() ? walkFiles(p) : [p];
  });
}
for (const file of [
  ...walkFiles(path.join(assets, 'cards/full')),
  ...walkFiles(path.join(assets, 'cards/thumb')),
  ...walkFiles(path.join(assets, 'boosters')),
]) {
  const size = fs.statSync(file).size;
  if (size > MAX_ASSET_BYTES) {
    warn('ASSET_TOO_HEAVY', 'Asset supérieur à 8 Mio.', {
      path: path.relative(assets, file).replaceAll(path.sep, '/'),
      bytes: size,
    });
  }
}

for (const [k, v] of Object.entries(ui)) {
  if (typeof v === 'string' && v.startsWith('data:image')) saveData(v, `ui/${slug(k)}`);
}

const idRegistry = Object.fromEntries(cards.map(card => [identityKey(card.setId, card.number), card.canonicalId]));
const validation = {
  schemaVersion: CATALOG_SCHEMA_VERSION,
  valid: errors.length === 0,
  summary: { cards: cards.length, extensions: extensions.length, errors: errors.length, warnings: warnings.length },
  errors,
  warnings,
};
const manifest = {
  schemaVersion: CATALOG_SCHEMA_VERSION,
  cardCount: cards.length,
  variantCount: cards.reduce((n, c) => n + c.variants.length, 0),
  extensionIds: extensions.map(e => e.id),
};

fs.writeFileSync(path.join(catalogDir, 'cards.json'), JSON.stringify(cards, null, 2));
fs.writeFileSync(path.join(catalogDir, 'extensions.json'), JSON.stringify(extensions, null, 2));
fs.writeFileSync(path.join(catalogDir, 'manifest.json'), JSON.stringify(manifest, null, 2));
fs.writeFileSync(path.join(catalogDir, 'id_registry.json'), JSON.stringify(idRegistry, null, 2));
fs.writeFileSync(path.join(catalogDir, 'validation_report.json'), JSON.stringify(validation, null, 2));

const inventory = [
  '# Asset Inventory', '',
  `- Schéma catalogue : **v${CATALOG_SCHEMA_VERSION}**`,
  `- Cartes canoniques : **${cards.length}**`,
  `- Variantes illustrées extraites : **${cards.reduce((n, c) => n + c.variants.length, 0)}**`,
  `- Boosters officiels : **${extensions.filter(e => e.boosterPath).length}**`,
  `- Erreurs de validation : **${errors.length}**`,
  `- Avertissements : **${warnings.length}**`,
  '', '## Extensions',
  ...extensions.map(e => `- ${e.name}: ${e.cardCount} cartes — \`${e.boosterPath || 'booster manquant'}\``),
  '', '## Identité technique',
  'Les IDs déjà publiés sont figés par `(extension, numéro)` afin qu’un renommage visible ne casse pas les sauvegardes. Les nouvelles cartes reçoivent un ID permanent de type `CARD_SET_NUMERO`.',
];
fs.writeFileSync(path.join(docs, 'ASSET_INVENTORY.md'), inventory.join('\n'));
fs.writeFileSync(path.join(docs, 'CARD_CATALOG.md'), '# Card Catalog\n\n' + cards.map(c => `- \`${c.canonicalId}\` — ${c.name} — ${c.rarity} — ${c.variants.length} variante(s)`).join('\n'));
fs.writeFileSync(path.join(docs, 'CONTENT_VALIDATION.md'), [
  '# Validation du contenu', '',
  `Statut : **${validation.valid ? 'VALIDE' : 'INVALIDE'}**`,
  `Erreurs : **${errors.length}** · Avertissements : **${warnings.length}**`, '',
  '## Erreurs',
  ...(errors.length ? errors.map(x => `- **${x.code}** — ${x.message} — \`${JSON.stringify(x.context)}\``) : ['- Aucune']),
  '', '## Avertissements',
  ...(warnings.length ? warnings.map(x => `- **${x.code}** — ${x.message} — \`${JSON.stringify(x.context)}\``) : ['- Aucun']),
].join('\n'));

console.log(JSON.stringify({
  schemaVersion: CATALOG_SCHEMA_VERSION,
  cards: cards.length,
  variants: cards.reduce((n, c) => n + c.variants.length, 0),
  extensions,
  validation: validation.summary,
}, null, 2));

if (errors.length) {
  console.error(`Catalogue invalide: ${errors.length} erreur(s), ${warnings.length} avertissement(s).`);
  process.exit(1);
}
